core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java

```diff
@@ -4044,6 +4049,52 @@ private OutputColumn createOutputColumn(Field field)
         {
             return new OutputColumn(new Column(field.getName().orElseThrow(), field.getType().toString()), analysis.getSourceColumns(field));
         }
+
+        /**
+         * Helper function that analyzes any versioning and returns the appropriate table handle.
+         * If no for clause exists, this is just a wrapper around getRedirectionAwareTableHandle in MetadataManager.
+         */
+        private RedirectionAwareTableHandle getTableHandle(Table table, QualifiedObjectName name, Optional<Scope> scope)
+        {
+            if (table.getQueryPeriod().isPresent()) {
+                Optional<TableVersion> tableVersion = extractTableVersion(table, scope);
+                Optional<TableVersionHandle> tableVersionHandle = metadata.getVersionHandle(session, name, tableVersion.get());
+                analysis.setTableVersion(table, tableVersionHandle.get());
+
+                return metadata.getRedirectionAwareTableHandle(session, name, tableVersionHandle);
+            }
+            return metadata.getRedirectionAwareTableHandle(session, name);
+        }
+
+        /**
+         * Analyzes a for clause and extracts an evaluated version value
+         */
+        private Optional<TableVersion> extractTableVersion(Table table, Optional<Scope> scope)
+        {
+            if (!table.getQueryPeriod().isPresent()) {
+                return Optional.empty();
+            }
+            Optional<TableVersion> tableVersion = Optional.empty();
+            Expression version = table.getQueryPeriod().get().getRangeExpression();
+
+            ExpressionAnalysis expressionAnalysis = analyzeExpression(version, scope.get());
+            analysis.recordSubqueries(table, expressionAnalysis);
+
+            // Once the range value is analyzed, we can evaluate it
+            Type versionType = expressionAnalysis.getType(version);
+            TravelType travelType = getTravelTypeFromRangeType(table.getQueryPeriod().get().getRangeType());
+            Object evaluatedVersion = evaluateConstantExpression(version, versionType, metadata, session, accessControl, ImmutableMap.of());
+            tableVersion = Optional.of(new TableVersion(travelType, versionType, evaluatedVersion));
+            return tableVersion;
+        }
+
+        private TravelType getTravelTypeFromRangeType(QueryPeriod.RangeType type)
+        {
+            if (type == QueryPeriod.RangeType.SYSTEM_TIME) {
+                return TravelType.TEMPORAL_TRAVEL;
+            }
+            return TravelType.VERSION_TRAVEL;
```

Comment:
Use a switch and fail if the value is unknown. As written, if a new QueryPeriod.RangeType is added to the system, it will fall through and return VERSION_TRAVEL, which is incorrect.

Reply:
Good point, cleaner implementation as well.